src/sql/src/plan/query.rs

```diff
@@ -2094,6 +2107,119 @@ fn plan_table_factor(
+        let mut outputs = Vec::with_capacity(scope.items.len());
+        // Re-jigger the outputs so the row number is first.
+        outputs.push(scope.items.len());
+        outputs.extend(0..scope.items.len());
+        let expr = HirRelationExpr::Project {
+            input: Box::new(HirRelationExpr::Map {
+                input: Box::new(HirRelationExpr::Join {
+                    left: Box::new(HirRelationExpr::Constant {
+                        rows: vec![Row::pack_slice(&[])],
+                        typ: RelationType {
+                            column_types: vec![],
+                            keys: vec![],
+                        },
+                    }),
+                    right: Box::new(expr),
+                    on: HirScalarExpr::Literal(
+                        Row::pack_slice(&[Datum::True]),
+                        ScalarType::Bool.nullable(false),
+                    ),
+                    kind: JoinKind::Inner { lateral: true },
+                }),
+                scalars: vec![HirScalarExpr::Windowing(WindowExpr {
+                    func: WindowExprType::Scalar(ScalarWindowExpr {
+                        func: ScalarWindowFunc::RowNumber,
+                        order_by: vec![],
+                    }),
+                    partition: vec![],
+                    order_by: vec![],
+                })],
+            }),
+            outputs,
+        };
+        last += 1;
+        // Use a Project to remove the row number columns.
+        project_outputs.extend(last..(last + scope.len()));
+        last += scope.len();
+        funcs.push((expr, scope));
+    }
+    let mut funcs = funcs.into_iter();
+
+    // Extract the first table function, which must exist.
+    let (mut expr, mut scope) = funcs.next().unwrap();
+
+    // Join remaining functions.
+    let mut last_column = 0;
+    let mut outputs = Vec::new();
+    for (idx, func) in funcs.enumerate() {
+        let (func, func_scope) = func;
+        let next_column = scope.items.len() + idx + 1;
+        expr = HirRelationExpr::Join {
+            left: Box::new(expr),
+            right: Box::new(func),
+            on: HirScalarExpr::CallBinary {
+                func: BinaryFunc::Eq,
+                expr1: Box::new(HirScalarExpr::Column(ColumnRef {
+                    level: 0,
+                    column: 0,
+                })),
+                expr2: Box::new(HirScalarExpr::Column(ColumnRef {
+                    level: 0,
+                    column: next_column,
+                })),
+            },
+            kind: JoinKind::FullOuter,
+        };
+        outputs.extend((last_column + 1)..next_column);
+        last_column = next_column;
+        scope.items.extend(func_scope.items);
+    }
+    let expr = HirRelationExpr::Project {
+        input: Box::new(expr),
+        outputs: project_outputs,
+    };
+    let scope = plan_table_function_alias(name, scope, alias)?;
+
+    Ok((expr, scope))
+}
```

Comment:
There's a bunch of nice builder functions on that'd be good to use, e.g. `.project` instead of manually constructing `HirRelationExpr::Project`, and `HirScalarExpr::literal_true()` instead of manually packing the literal. But otherwise all seems to make sense!